server/src/child.rs

```diff
@@ -502,18 +531,57 @@ fn call_function(
 
 fn trace_function(
     target: &mut TargetState, trace: &mut TraceState, tx: &SyncSender<DebugMessage>,
-    execution: ExecutionState
+    rx: &Receiver<ServerMessage>, execution: ExecutionState
 ) -> io::Result<()> {
     let (breakpoints, line, exit, call) = match execution {
         ExecutionState::Function { trace, line, exit, call } => (trace, line, exit, call),
         _ => unreachable!(),
     };
+
     let mut breakpoints = Trace::resume(&mut target.child, breakpoints);
 
     let mut last_line = line;
     let mut last_breakpoint = None;
     loop {
         let mut event = debug::Event::wait_event()?;
+        
+        match rx.recv().unwrap() {
+            ServerMessage::Stop => {
+                let TargetState { ref symbols, ref threads, .. } = *target;
+                let thread = threads[&event.thread_id];
+
+                let TraceState { event: ref mut trace_event, .. } = *trace;
+                *trace_event = Some(event);
+
+                let context = debug::get_thread_context(thread, winapi::CONTEXT_FULL)?;
+
+                // restart the instruction
+                //context.set_instruction_pointer(address);
+
+                // collect the return value
+                let (_, restore) = call.teardown(breakpoints.child(), &context, &symbols)?;
```

Comment:
There is no return value to collect if you're stopping in the middle of a function, so it would probably be better to add a new method to `Call` (maybe called `cancel`?) that just consumes `self` and returns `restore`.